flang/lib/Lower/OpenMP/DataSharingProcessor.cpp

```diff
@@ -116,6 +116,31 @@ void DataSharingProcessor::cloneSymbol(const semantics::Symbol *sym) {
       *sym, /*skipDefaultInit=*/isFirstPrivate);
   (void)success;
   assert(success && "Privatization failed due to existing binding");
+
+  // Initialize clone from original object if it has any allocatable member.
+  auto needInitClone = [&] {
+    if (isFirstPrivate)
+      return false;
+
+    SymbolBox sb = symTable.lookupSymbol(sym);
+    assert(sb);
+    mlir::Value addr = sb.getAddr();
+    assert(addr);
+    mlir::Type ty = addr.getType();
+
+    // Unwrap type
+    while (auto eleTy = fir::dyn_cast_ptrOrBoxEleTy(ty))
+      ty = eleTy;
+    // For arrays, use its element type
+    if (auto seqTy = mlir::dyn_cast<fir::SequenceType>(ty))
+      ty = seqTy.getEleTy();
+    return fir::isRecordWithAllocatableMember(ty);
```

Comment:
Note that this will also flag polymorphic objects as potentially having allocatable components (the dynamic type extending the static type may have such components and is unknown).

Reply:
Thanks for catching that.